services/blockstorage/block_sync_server.go

```diff
@@ -104,31 +105,55 @@ func (s *Service) sourceHandleBlockSyncRequest(ctx context.Context, message *gos
 	}
 
-	blocks, firstAvailableBlockHeight, lastAvailableBlockHeight, err := s.GetBlockSlice(firstRequestedBlockHeight, lastRequestedBlockHeight)
+	blocks, firstAvailableBlockHeight, _, err := s.GetBlockSlice(firstRequestedBlockHeight, lastRequestedBlockHeight)
 	if err != nil {
 		return errors.Wrap(err, "block sync failed reading from block persistence")
 	}
 
-	logger.Info("sending blocks to another node via block sync",
-		log.Stringable("petitioner", senderNodeAddress),
-		log.Uint64("first-available-block-height", uint64(firstAvailableBlockHeight)),
-		log.Uint64("last-available-block-height", uint64(lastAvailableBlockHeight)))
+	var chunkSize = len(blocks)
+	var lastError error
+
+	const maxAttempts = 100 // we will never server more than 2^100 blocks..
+	for attempt := 0; attempt < maxAttempts; attempt++ {
+		lastAvailableBlockHeight := firstAvailableBlockHeight + primitives.BlockHeight(chunkSize) - 1
+		logger.Info("sending blocks to another node via block sync",
+			log.Stringable("petitioner", senderNodeAddress),
+			log.Uint64("first-available-block-height", uint64(firstAvailableBlockHeight)),
+			log.Uint64("last-available-block-height", uint64(lastAvailableBlockHeight)))
+
+		response := &gossiptopics.BlockSyncResponseInput{
+			RecipientNodeAddress: senderNodeAddress,
+			Message: &gossipmessages.BlockSyncResponseMessage{
+				Sender: (&gossipmessages.SenderSignatureBuilder{
+					SenderNodeAddress: s.config.NodeAddress(),
+				}).Build(),
+				SignedChunkRange: (&gossipmessages.BlockSyncRangeBuilder{
+					BlockType:                blockType,
+					FirstBlockHeight:         firstAvailableBlockHeight,
+					LastBlockHeight:          lastAvailableBlockHeight,
+					LastCommittedBlockHeight: lastCommittedBlockHeight,
+				}).Build(),
+				BlockPairs: blocks[:chunkSize],
+			},
+		}
+		_, err = s.gossip.SendBlockSyncResponse(ctx, response)
+		if err == nil {
+			return nil
+		}
+		if !gossip.IsChunkTooBigError(err) {
+			return err
+		}
+		lastError = err
+
+		if chunkSize == 0 { // test before dividing by 2, to attempt a zero length chunk at least once
+			break
+		}
+		chunkSize /= 2 // Chunk too big, retry with a smaller chunk
+	}
 
-	response := &gossiptopics.BlockSyncResponseInput{
-		RecipientNodeAddress: senderNodeAddress,
-		Message: &gossipmessages.BlockSyncResponseMessage{
-			Sender: (&gossipmessages.SenderSignatureBuilder{
-				SenderNodeAddress: s.config.NodeAddress(),
-			}).Build(),
-			SignedChunkRange: (&gossipmessages.BlockSyncRangeBuilder{
-				BlockType:                blockType,
-				FirstBlockHeight:         firstAvailableBlockHeight,
-				LastBlockHeight:          lastAvailableBlockHeight,
-				LastCommittedBlockHeight: lastCommittedBlockHeight,
-			}).Build(),
-			BlockPairs: blocks,
-		},
+	// Must be an error indicating chunk-too-big
+	if !gossip.IsChunkTooBigError(lastError) {
+		panic(errors.Wrap(err, "unreachable code - expected chunk-too-big error"))
```

Comment:
this is one of these ironic errors you end up finding in logs. You actually panic the goroutine here - is this what you intended to do?

Reply:
Yes - this is an assertion, if it actually executes then something in the state is fundamentally broken